wasm-chain-integration/src/v1/trie/low_level.rs

```diff
@@ -2331,3 +2343,154 @@ impl<V: Clone> Node<V> {
         true
     }
 }
+
+enum SubtreeAction {
+    InsertOrDelete,
+    DeletePrefix,
+}
+/// todo doc.
+fn is_subtree_locked(
+    action: SubtreeAction,
+    iterator_roots: &std::collections::BTreeMap<Vec<u8>, u16>,
+    key: &[u8],
+) -> bool {
+    for iter_root in iterator_roots.keys() {
+        match action {
+            SubtreeAction::InsertOrDelete => {
+                if key.starts_with(iter_root) {
+                    return true;
+                }
+            }
+            SubtreeAction::DeletePrefix => {
+                if key.starts_with(iter_root) || iter_root.starts_with(key) {
+                    return true;
+                }
+            }
+        }
+    }
+    false
+}
+
+/// Lock the subtree for modifications
+/// The key is the root of the subtree to be locked.
+/// Returns true if the iterator was added to the iterator_roots otherwise false
+fn lock_subtree(iterator_roots: &mut std::collections::BTreeMap<Vec<u8>, u16>, key: &[u8]) -> bool {
+    if let Some(iter) = iterator_roots.get_mut(key) {
+        if let Some(new_value) = iter.checked_add(1) {
+            *iter = new_value;
+            true
+        } else {
+            false
+        }
+    } else {
+        iterator_roots.insert(key.to_vec(), 1);
+        true
+    }
+}
+
+/// Unlock the tree for modifications.
+/// PRECONDITION: The provided key is an actual root of an iterator.
+fn unlock_subtree(
+    iterator_roots: &mut std::collections::BTreeMap<Vec<u8>, u16>,
+    key: &[u8],
+) -> bool {
+    if let Some(rc) = iterator_roots.get_mut(key) {
+        *rc = rc.saturating_sub(1);
+        if *rc == 0 {
+            iterator_roots.remove(key);
+        }
+        true
+    } else {
+        false
+    }
+}
+
+/// todo write these tests with quickcheck
+#[test]
+fn test_locking_insert() -> anyhow::Result<()> {
```

Comment:
This must go into tests.rs

Reply:
Then I would have to expose the functions - would rather not since they are an implementation detail of the tree.

Reply:
Then it must at least go into a separate module.

This is how we do test elsewhere, not just inline.